### ISSUE TITLE
[Formula] which functions to add (needed by geocachers)

### ISSUE THREAD

**Issue by eddiemuc** [cgeo/cgeo#12386]
This issue should collect functions that should be added to formula which are commonly needed by Geocachers. Implementation already contains "length",  "checksum"/"lettervalue" and "rot". What else is needed out there?

Out of the top of my hat:
* roman numerals (enter e.g. "xii", get 12, or vice versa)
* handycode (what is the englisch word? Phonecode?) (enter "cache", get 22243)
* morse (enter "... --- ...", get "SOS", or vice versa)

I would restrain from entering too exotic or computation intensive functions. Also, functions must be constant (same input values always lead to same result) due to caching mechanism inside Formula

Any more ideas what could be important (or maybe concerns NOT to include some things)?

**Comment by eddiemuc** [comment]
And BTW: happy christmas everyone!

**Comment by Lineflyer** [comment]
> (what is the englisch word? Phonecode?) 

Vanity Code

**Comment by Lineflyer** [comment]
Regarding functions to add:
Theoretically many features/funtions would be possible, but maybe not meaningful as only rarely been used.

Lets look at this from the use case perspective:
IMHO the variable tab will be mostly used, while preparing a geocaching trip and/or when it is known what type of conversion is needed to get the desired result.
This means to me, that it does not make sense to add complex functions e.g. for various ciphers, as usually you will only find the ciphered value while being on-site and finding the correct deciphering method is part of the field puzzle but not provided in advance.

When focusing on what is usually provided by the owner in the listing, I would see:
- [x] Checksum
- [x] Iterative checksum
- [x] Letter values
- [ ] Roman -> arabic numerals (Remark: Why would you need arabic -> roman. Can/Shall the roman value be used for calculation?)

Vanity code and morse are IMHO already ciphers, which are not necessarily needed. I would suggest to concentrate on on making the basic conversions listed above very easy to use, instead of putting more functions in, which are better suited in some special app (like GC wizard) anyway.

**Comment by SammysHP** [comment]
Hard to say. I'd include only mathematical functions (square root etc).

Would be nice to have an API for GC Wizard so that we can call its functions and get the calculated value as a response.

1. User selects "calculate with GC Wizard"
2. GC Wizard opens and provides a list of available functions
3. User selects function, optionally mandatory "special" options
4. GC Wizard returns a configuration string to c:geo containing
  - input variables and type
  - return type
  - name of function
  - some ID or encoded function that can be sent to GC Wizard to recall the function
5. c:geo adds this data to a calculation step/variable
6. User enters input variables
7. c:geo calls the function in the background with the stored information whenever it is necessary

@S-Man42 What do you think, does this sound possible?

**Comment by S-Man42** [comment]
Hi,

(as @SammysHP already edited) there are several thoughts about such things as a Public API. https://github.com/S-Man42/GCWizard/issues/489

We'd be very proud of such an amazing partnership and we also believe, that it wouldn't make any sense to develop such functions in parallel.

However, tbh, I have absolutely no idea of inter-app communication. We would love to provide anything you need, but I guess, we need to find to solve the technical question of this issues: How to provide an interface between two apps (especially the on our Flutter-based side, we are currently without any skills), how should a good api look like and what functions do you need.

I am sure, this would be very useful for another use case, @capoaira was working on: Provide a public interface in the web version to combine it with websites or greasemonkey scripts. https://github.com/S-Man42/GCWizard/issues/488

Any thoughts and skills on that are appreciated.

**Comment by eddiemuc** [comment]
> Hard to say. I'd include only mathematical functions (square root etc).
> 
> Would be nice to have an API for GC Wizard so that we can call its functions and get the calculated value as a response.

This is a great idea, but in order for it to work performance-wise we would need the functions as a library to include into c:geo. Using an API would be a performance-desaster since functions are constantly re-calculated on-the-fly while the user types. I don't know much about app-to-app-communication, but knowing Android for a while now I assume this is intent-based and asynchronous.

Would such a lib be available?

**Comment by ztNFny** [comment]
Imho the coord calculator in c:geo should stay just that - a tool to find coordinates.
More advanced calculations like letter word, cross sum, ... should stay in a separate tool. Otherwise where do you draw the line, what don't you implement? If we can interface with GCWizard for that "advanced" stuff great!

**Comment by moving-bits** [comment]
I guess it will be hard to define a specific line which functions to include, and which not. Also, at least when you have to type (and probably: to remember) the functions names, it may be easier to open a specific app for that, especially as they may give you a formula-specific UI for entering the formula's parameters. But some basic functions would be nice to have.

For me most frequently used functions in a geocaching context would be checksum, iterative checksum and word (letter) value, beside some basic mathematical functions.

**Comment by eddiemuc** [comment]
From what I read in the comments I would suggest to add vintage code and Roman numbers and leave it at that. Then we would have the same set of functions as GCBuddy, and I guess this makes sense.

Vintage code I needed in the field just yesterday

**Comment by SammysHP** [comment]
I vote against adding any conversion/encoding functions directly. Why vanity and Roman numbers and not also character value (A=1, B=2...)? Let's keep it purely mathematical and use other apps for everything else.

**Comment by fm-sys** [comment]
> Why vanity and Roman numbers and not also character value

Letter value _is_ already implemented BTW...

**Comment by ztNFny** [comment]
Cross sum is ok i guess as it's a mathematical function. Char values maybe ok as well as it's the most common non-calculation.
Roman and vanity are rarely used (imho) - but the problem here is that it's kinda Impossible to draw the line

**Comment by eddiemuc** [comment]
> Roman and vanity are rarely used (imho) - but the problem here is that it's kinda Impossible to draw the line

My suggestion for such a line: offer same set of functions as GCBuddy. Adding vanity and Roman puts us there.

Without some basic function set the whole variables don't make much sense imho

**Comment by eddiemuc** [comment]
> Let's keep it purely mathematical and use other apps for everything else.

I don't understand why "mathematical" should be the defining set. We have e.g. implemented trigonometric functions (sin, cos, tan), this was already there before my doing and I left it there of course. It doesn't hurt either. But I really never needed this any time, and it is most definitely needed less than e.g. word value, vanity or roman (or Morse for that matter)

**Comment by ztNFny** [comment]
> We have e.g. implemented trigonometric functions (sin, cos, tan), this was already there before my doing and I left it there of course. It doesn't hurt either. 

Imho we could remove this to reduce code complexity. A little less code to maintain doesnt hurt and this is really not something common.

I'm.ok with using GCBuddy as "the line".

**Comment by eddiemuc** [comment]
Created separate issue for GC Wizard #12389. 
PR #12388 will include vanity and roman into Formula.

As of now this seems a reasonable solution to this issue to me.
